File: api/internal/promoter/event/timetable.go
package event

import (
	"sort"
	"time"

	"github.com/google/uuid"
)

// slot is one timed performance on a stage; a b2b group is a single slot.
type slot struct {
	start, end time.Time
	entries    []uuid.UUID
}

// ValidateTimetable checks the lineup against the event window and each
// stage's curfew and changeover. It is the single source of truth: the UI
// mirrors it for instant feedback, but only this result gates release.
func ValidateTimetable(ev Event, stages []Stage, lineup []LineupEntry) []Issue {
	var issues []Issue
	byID := map[uuid.UUID]Stage{}
	for _, s := range stages {
		byID[s.ID] = s
	}
	windowStart := ev.StartsAt
	if ev.DoorsAt != nil {
		windowStart = *ev.DoorsAt
	}

	perStage := map[uuid.UUID][]slot{}
	type groupKey struct {
		stage uuid.UUID
		group int
	}
	groups := map[groupKey]*slot{}
	var order []groupKey

	for _, e := range lineup {
		if e.SetStart == nil || e.SetEnd == nil {
			issues = append(issues, Issue{Code: "untimed", Severity: SeverityWarning, EntryIDs: []uuid.UUID{e.ID}})
			continue
		}
		if e.StageID == nil {
			issues = append(issues, Issue{Code: "no_stage", Severity: SeverityError, EntryIDs: []uuid.UUID{e.ID}})
			continue
		}
		st, ok := byID[*e.StageID]
		if !ok {
			issues = append(issues, Issue{Code: "unknown_stage", Severity: SeverityError, EntryIDs: []uuid.UUID{e.ID}})
			continue
		}
		start, end := *e.SetStart, *e.SetEnd
		if start.Before(windowStart) || end.After(ev.EndsAt) {
			issues = append(issues, Issue{Code: "outside_event", Severity: SeverityError, StageID: &st.ID,
				EntryIDs: []uuid.UUID{e.ID}, From: &start, To: &end})
		}
		if st.CurfewAt != nil && end.After(*st.CurfewAt) {
			curfew := *st.CurfewAt
			issues = append(issues, Issue{Code: "past_curfew", Severity: SeverityError, StageID: &st.ID,
				EntryIDs: []uuid.UUID{e.ID}, From: &curfew, To: &end, Minutes: int(end.Sub(curfew).Minutes())})
		}
		if e.B2BGroup != nil {
			k := groupKey{st.ID, *e.B2BGroup}
			if g, ok := groups[k]; ok {
				if !g.start.Equal(start) || !g.end.Equal(end) {
					issues = append(issues, Issue{Code: "b2b_mismatch", Severity: SeverityError, StageID: &st.ID,
						EntryIDs: append(append([]uuid.UUID{}, g.entries...), e.ID)})
				}
				g.entries = append(g.entries, e.ID)
				continue
			}
			groups[k] = &slot{start: start, end: end, entries: []uuid.UUID{e.ID}}
			order = append(order, k)
			continue
		}
		perStage[st.ID] = append(perStage[st.ID], slot{start: start, end: end, entries: []uuid.UUID{e.ID}})
	}
	for _, k := range order {
		perStage[k.stage] = append(perStage[k.stage], *groups[k])
	}

	stageIDs := make([]uuid.UUID, 0, len(perStage))
	for id := range perStage {
		stageIDs = append(stageIDs, id)
	}
	sort.Slice(stageIDs, func(i, j int) bool { return byID[stageIDs[i]].Position < byID[stageIDs[j]].Position })

	for _, id := range stageIDs {
		st := byID[id]
		slots := perStage[id]
		sort.Slice(slots, func(i, j int) bool { return slots[i].start.Before(slots[j].start) })
		changeover := time.Duration(st.ChangeoverMinutes) * time.Minute
		for i := 1; i < len(slots); i++ {
			prev, next := slots[i-1], slots[i]
			stageID := st.ID
			entries := append(append([]uuid.UUID{}, prev.entries...), next.entries...)
			gap := next.start.Sub(prev.end)
			switch {
			case gap < 0:
				from, to := next.start, prev.end
				issues = append(issues, Issue{Code: "overlap", Severity: SeverityError, StageID: &stageID,
					EntryIDs: entries, From: &from, To: &to, Minutes: int(-gap.Minutes())})
			case gap < changeover:
				issues = append(issues, Issue{Code: "short_changeover", Severity: SeverityWarning, StageID: &stageID,
					EntryIDs: entries, Minutes: int(gap.Minutes())})
			case gap > changeover:
				from, to := prev.end.Add(changeover), next.start
				issues = append(issues, Issue{Code: "dead_air", Severity: SeverityWarning, StageID: &stageID,
					From: &from, To: &to, Minutes: int(to.Sub(from).Minutes())})
			}
		}
	}
	return issues
}

File: apps/promoter/app/utils/timetable.ts
import type { EventDetail, Issue, LineupEntry, Stage } from '~/types/event'

/**
 * Client mirror of api/internal/promoter/event/timetable.go for instant
 * feedback while editing. The server result is authoritative; parity is
 * pinned by utils/__tests__/timetable.test.ts (same cases as the Go tests).
 */
type Window = Pick<EventDetail, 'starts_at' | 'ends_at' | 'doors_at'>

interface Slot { start: number, end: number, entries: string[] }

const ms = (iso: string) => Date.parse(iso)

export function validateTimetable(ev: Window, stages: Stage[], lineup: LineupEntry[]): Issue[] {
  const issues: Issue[] = []
  const byId = new Map(stages.map(s => [s.id, s]))
  const windowStart = ms(ev.doors_at ?? ev.starts_at)
  const windowEnd = ms(ev.ends_at)
  const perStage = new Map<string, Slot[]>()
  const groups = new Map<string, Slot & { stage: string }>()

  for (const e of lineup) {
    if (!e.set_start || !e.set_end) {
      issues.push({ code: 'untimed', severity: 'warning', entry_ids: [e.id] })
      continue
    }
    if (!e.stage_id) {
      issues.push({ code: 'no_stage', severity: 'error', entry_ids: [e.id] })
      continue
    }
    const st = byId.get(e.stage_id)
    if (!st) {
      issues.push({ code: 'unknown_stage', severity: 'error', entry_ids: [e.id] })
      continue
    }
    const start = ms(e.set_start)
    const end = ms(e.set_end)
    if (start < windowStart || end > windowEnd) {
      issues.push({ code: 'outside_event', severity: 'error', stage_id: st.id, entry_ids: [e.id], from: e.set_start, to: e.set_end })
    }
    if (st.curfew_at && end > ms(st.curfew_at)) {
      issues.push({ code: 'past_curfew', severity: 'error', stage_id: st.id, entry_ids: [e.id], from: st.curfew_at, to: e.set_end, minutes: Math.round((end - ms(st.curfew_at)) / 60000) })
    }
    if (e.b2b_group != null) {
      const key = `${st.id}:${e.b2b_group}`
      const g = groups.get(key)
      if (g) {
        if (g.start !== start || g.end !== end) {
          issues.push({ code: 'b2b_mismatch', severity: 'error', stage_id: st.id, entry_ids: [...g.entries, e.id] })
        }
        g.entries.push(e.id)
        continue
      }
      groups.set(key, { start, end, entries: [e.id], stage: st.id })
      continue
    }
    perStage.set(st.id, [...(perStage.get(st.id) ?? []), { start, end, entries: [e.id] }])
  }
  for (const g of groups.values()) perStage.set(g.stage, [...(perStage.get(g.stage) ?? []), g])

  const ids = [...perStage.keys()].sort((a, b) => (byId.get(a)!.position - byId.get(b)!.position))
  for (const id of ids) {
    const st = byId.get(id)!
    const slots = perStage.get(id)!.sort((a, b) => a.start - b.start)
    const changeover = st.changeover_minutes * 60000
    // Compare with the latest end so far (a short set inside a long one).
    let last = slots[0]!
    for (let i = 1; i < slots.length; i++) {
      const next = slots[i]!
      const gap = next.start - last.end
      const entries = [...last.entries, ...next.entries]
      if (gap < 0) {
        const to = Math.min(last.end, next.end)
        issues.push({ code: 'overlap', severity: 'error', stage_id: id, entry_ids: entries, from: new Date(next.start).toISOString(), to: new Date(to).toISOString(), minutes: Math.round((to - next.start) / 60000) })
      } else if (gap < changeover) {
        issues.push({ code: 'short_changeover', severity: 'warning', stage_id: id, entry_ids: entries, minutes: Math.round(gap / 60000) })
      } else if (gap > changeover) {
        const from = last.end + changeover
        issues.push({ code: 'dead_air', severity: 'warning', stage_id: id, from: new Date(from).toISOString(), to: new Date(next.start).toISOString(), minutes: Math.round((next.start - from) / 60000) })
      }
      if (next.end > last.end) last = next
    }
  }
  return issues
}

export const hasErrors = (issues: Issue[]) => issues.some(i => i.severity === 'error')
